lib/tracer/tracer.js

```diff
@@ -58,29 +69,37 @@ Tracer.prototype.start = function (name, type, {
     events: [],
     userId,
   };
-
-  return traceInfo;
 };
 
-Tracer.prototype.event = function (traceInfo, type, data, metaData) {
-  // do not allow to proceed, if already completed or errored
-  let lastEvent = this.getLastEvent(traceInfo);
+Tracer.prototype.event = function (trace, type, data, meta) {
+  const activeEvent = getActiveEvent();
 
+  // We should not nest based on the last event,
+  // but based on the active event of the current context, due to parallel execution.
+  const lastEvent = trace.events[trace.events.length - 1];
+
+  // Do not allow proceeding, if already completed or errored
   if (
-    // trace completed but has not been processed
-    lastEvent &&
-    ['complete', 'error'].indexOf(lastEvent.type) >= 0 ||
-    // trace completed and processed.
-    traceInfo.isEventsProcessed
+    // Trace completed but has not been processed
+    [EventType.Complete, EventType.Error].includes(lastEvent?.type) ||
+    // Trace completed and processed.
+    trace.isEventsProcessed
   ) {
     return false;
   }
 
+  if (type === EventType.Start) {
+    trace.rootAsyncId = executionAsyncId();
+  }
+
   let event = {
     type,
     at: Ntp._now(),
     endAt: null,
     nested: [],
+    asyncId: executionAsyncId(),
+    triggerAsyncId: triggerAsyncId(),
+    ...meta,
```

Comment:
For personal preference, I preferred how it was explicit which properties from meta were copied to the event. To me it makes it easier to maintain and understand the shape of the event object.